File: frontend/src/types/course.ts
export type Course = {
  id: string;
  title: string;
  description: string;
  instructorName: string;
  thumbnailUrl?: string;
  level?: number | string;
  category?: string;
  tags?: string[];
  duration?: string;
};

export type CourseStatus = 'draft' | 'published' | 'hidden' | 'archived';

export type CourseRecord = {
  id: number;
  title: string;
  description: string | null;
  level: string;
  category: string;
  learning_outcomes: string[];
  requirements: string[];
  thumbnail_url: string | null;
  price: number;
  status: CourseStatus;
  instructor_id: number;
  created_at: string;
  updated_at: string;
};

export type LessonRecord = {
  id: number;
  course_id: number;
  title: string;
  content: string | null;
  video_url: string | null;
  material_url: string | null;
  order_index: number;
};

export type LessonPayload = {
  title: string;
  content?: string;
  video_url?: string;
  material_url?: string;
  order_index: number;
};

export type LessonUpdatePayload = Partial<LessonPayload>;

export type CourseDetailRecord = CourseRecord & {
  lessons: LessonRecord[];
};

export type CourseCreatePayload = {
  title: string;
  description?: string;
  level: string;
  category: string;
  learning_outcomes: string[];
  requirements: string[];
  thumbnail_url?: string;
  price: number;
  status: Exclude<CourseStatus, 'archived'>;
};

export type CourseUpdatePayload = Partial<CourseCreatePayload>;

File: backend/app/models/course.py
from sqlalchemy import JSON, Column, DateTime, ForeignKey, Integer, Numeric, String, Text
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from app.database import Base


class Course(Base):
    __tablename__ = "courses"

    id = Column(Integer, primary_key=True, index=True)
    title = Column(String, nullable=False)
    description = Column(Text, nullable=True)
    level = Column(String(50), nullable=False, default="All levels")
    category = Column(String(100), nullable=False, default="EduCloud")
    learning_outcomes = Column(JSON, nullable=False, default=list)
    requirements = Column(JSON, nullable=False, default=list)
    thumbnail_url = Column(String, nullable=True)
    price = Column(Numeric(10, 2), nullable=False, default=0)
    status = Column(String, nullable=False, default="draft")
    instructor_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())

    lessons = relationship(
        "Lesson",
        back_populates="course",
        cascade="all, delete-orphan",
        order_by="Lesson.order_index",
    )
    assessment = relationship(
        "CourseAssessment",
        back_populates="course",
        cascade="all, delete-orphan",
        uselist=False,
    )
